pkg/apis/sessions/session_state.go

```diff
@@ -11,22 +16,24 @@ import (
 )
 
 // SessionState is used to store information about the currently authenticated user session
+// MessagePack naming aligned to match with internal JWT claims for compression synergy
 type SessionState struct {
-	AccessToken       string    `json:",omitempty"`
-	IDToken           string    `json:",omitempty"`
-	CreatedAt         time.Time `json:"-"`
-	ExpiresOn         time.Time `json:"-"`
-	RefreshToken      string    `json:",omitempty"`
-	Email             string    `json:",omitempty"`
-	User              string    `json:",omitempty"`
-	PreferredUsername string    `json:",omitempty"`
+	AccessToken       string    `json:",omitempty" msgpack:"at,omitempty"`
+	IDToken           string    `json:",omitempty" msgpack:"it,omitempty"`
+	CreatedAt         time.Time `json:"-" msgpack:"-"`
+	ExpiresOn         time.Time `json:"-" msgpack:"-"`
+	RefreshToken      string    `json:",omitempty" msgpack:"rt,omitempty"`
+	Email             string    `json:",omitempty" msgpack:"e,omitempty"`
+	User              string    `json:",omitempty" msgpack:"u,omitempty"`
+	PreferredUsername string    `json:",omitempty" msgpack:"pu,omitempty"`
 }
 
-// SessionStateJSON is used to encode SessionState into JSON without exposing time.Time zero value
-type SessionStateJSON struct {
+// SessionStateEncoded is used to encode SessionState into JSON/MessagePack
+// without exposing time.Time zero value
+type SessionStateEncoded struct {
 	*SessionState
-	CreatedAt *time.Time `json:",omitempty"`
-	ExpiresOn *time.Time `json:",omitempty"`
+	CreatedAt *time.Time `json:",omitempty" msgpack:"ca,omitempty"`
+	ExpiresOn *time.Time `json:",omitempty" msgpack:"eo,omitempty"`
```

Comment:
I realise you didn't add it, but do you think we could merge these two? ie always have time pointers? Would that still work?

Reply:
Gave this a whirl in this commit: https://github.com/oauth2-proxy/oauth2-proxy/pull/523/commits/124eb5b854d89f6fd9bb82482c215d739c785b70

I didn't know why the special time logic -- but I left it intact to not reintroduce whatever bug it might've fixed in the past.